test/framework/toolchain.py

```diff
@@ -381,6 +381,42 @@ def test_optarch_aarch64_heuristic(self):
         self.assertEqual(tc.options.options_map['optarch'], 'mcpu=cortex-a72.cortex-a53')
         self.assertTrue('-mcpu=cortex-a72.cortex-a53' in os.environ['CFLAGS'])
         self.modtool.purge()
+  
+    def test_compiler_dependent_optarch(self):
+        """Test whether specifying optarch on a per compiler basis works."""
+        flag_vars = ['CFLAGS', 'CXXFLAGS', 'FCFLAGS', 'FFLAGS', 'F90FLAGS']
+        for intelflags,intel_expanded_flags in [('intelflag','intelflag'),('GENERIC','xSSE2'),('','')]:
+            for gccflags,gcc_expanded_flags in  [('gccflag','gccflag'),('GENERIC','march=x86-64 -mtune=generic'),('','')]:
+                optarch_var = 'Intel:%s,GCC:%s' % (intelflags,gccflags)
+                build_options = {'optarch': optarch_var}
+                init_config(build_options=build_options)
+                for toolchain, toolchain_version in [('iccifort','2011.13.367'),('GCC','4.7.2'),('PGI','16.7-GCC-5.4.0-2.26')]:
+                    for enable in [True, False]:
+               	        tc = self.get_toolchain(toolchain, version=toolchain_version)
+                        tc.set_options({'optarch': enable})
+                        tc.prepare()
+                        flag = None
+                        if toolchain == 'iccifort':
+                            flag = intel_expanded_flags
+                        elif toolchain == 'GCC':
+                            flag = gcc_expanded_flags 
+                        else: # PGI as an example of compiler not set
+                            # default optarch flag
+                            flag = tc.COMPILER_OPTIMAL_ARCHITECTURE_OPTION[(tc.arch,tc.cpu_family)]
```

Comment:
This should be handled outside the for loop, where you construct a list of test cases?

Also, hardcode the value for PGI rather than grabbing it from the `Toolchain` instance?